File: sections/41_euler_tour_lca_binary_lifting/problems/b_max_edge_path/solution.py
import sys


def main():
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    n, q = data[0], data[1]
    idx = 2
    LOG = max(1, (n + 1).bit_length())

    up = [[0] * (n + 1) for _ in range(LOG)]
    mx = [[0] * (n + 1) for _ in range(LOG)]
    depth = [0] * (n + 1)
    for v in range(2, n + 1):
        p, w = data[idx], data[idx + 1]
        idx += 2
        up[0][v] = p
        mx[0][v] = w
        depth[v] = depth[p] + 1
    for k in range(1, LOG):
        upp, upc = up[k - 1], up[k]
        mxp, mxc = mx[k - 1], mx[k]
        for v in range(1, n + 1):
            mid = upp[v]
            upc[v] = upp[mid]
            mxc[v] = max(mxp[v], mxp[mid])

    def lift(v, d):
        """Raise v by d; return (vertex, heaviest edge crossed)."""
        best = 0
        bit = 0
        while d:
            if d & 1:
                best = max(best, mx[bit][v])
                v = up[bit][v]
            d >>= 1
            bit += 1
        return v, best

    out = []
    for _ in range(q):
        a, b = data[idx], data[idx + 1]
        idx += 2
        if depth[a] < depth[b]:
            a, b = b, a
        a, ans = lift(a, depth[a] - depth[b])
        if a != b:
            for k in range(LOG - 1, -1, -1):
                if up[k][a] != up[k][b]:
                    ans = max(ans, mx[k][a], mx[k][b])
                    a = up[k][a]
                    b = up[k][b]
            ans = max(ans, mx[0][a], mx[0][b])
        out.append(str(ans))
    print("\n".join(out))


if __name__ == "__main__":
    main()

File: sections/41_euler_tour_lca_binary_lifting/problems/b_max_edge_path/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, q;
    if (!(cin >> n >> q)) return 0;

    int LOG = 1;
    while ((1 << LOG) <= n + 1) LOG++;

    vector<vector<int>> up(LOG, vector<int>(n + 1));
    vector<vector<int>> mx(LOG, vector<int>(n + 1));
    vector<int> depth(n + 1);

    for (int v = 2; v <= n; v++) {
        int p, w;
        cin >> p >> w;
        up[0][v] = p;
        mx[0][v] = w;
        depth[v] = depth[p] + 1;
    }
    for (int k = 1; k < LOG; k++)
        for (int v = 1; v <= n; v++) {
            int mid = up[k - 1][v];
            up[k][v] = up[k - 1][mid];
            mx[k][v] = max(mx[k - 1][v], mx[k - 1][mid]);
        }

    // Raises v by d edges and returns the heaviest edge crossed.
    auto lift = [&](int &v, int d) {
        int best = 0;
        for (int k = 0; k < LOG; k++)
            if (d & (1 << k)) {
                best = max(best, mx[k][v]);
                v = up[k][v];
            }
        return best;
    };

    while (q--) {
        int a, b;
        cin >> a >> b;
        if (depth[a] < depth[b]) swap(a, b);
        int ans = lift(a, depth[a] - depth[b]);
        if (a != b) {
            for (int k = LOG - 1; k >= 0; k--)
                if (up[k][a] != up[k][b]) {
                    ans = max({ans, mx[k][a], mx[k][b]});
                    a = up[k][a];
                    b = up[k][b];
                }
            ans = max({ans, mx[0][a], mx[0][b]});
        }
        cout << ans << '\n';
    }
}
